### ISSUE TITLE
Stackoverflow with u8 literal concatenation on macOS

### ISSUE THREAD

**Issue by am11** [dotnet/roslyn#62361]
**Version Used**: 4.4.0-1.22328.22

**Steps to Reproduce**:
[jcouv update:] This is not specific to MacOS.

Compile with `dotnet build -p:LangVersion=preview`:
```c#
internal static class X
{
  public  static ReadOnlySpan<byte> Y =>
    "a"u8 +
    "a"u8 +
// repeat until Y.Length is 741 (or higher; 740 is somehow the limit the compiler can withstand on macs)
    "a"u8 +
    "a"u8;
}
```

**Expected Behavior**:

Compilation succeed.

**Actual Behavior**:

```
/Users/am11/.dotnet7/sdk/7.0.100-preview.7.22351.3/Roslyn/Microsoft.CSharp.Core.targets(75,5): error : Stack overflow. [/Users/am11/nativeapp1/nativeapp1.csproj]
/Users/am11/.dotnet7/sdk/7.0.100-preview.7.22351.3/Roslyn/Microsoft.CSharp.Core.targets(75,5): error : Repeat 741 times: [/Users/am11/nativeapp1/nativeapp1.csproj]
/Users/am11/.dotnet7/sdk/7.0.100-preview.7.22351.3/Roslyn/Microsoft.CSharp.Core.targets(75,5): error : -------------------------------- [/Users/am11/nativeapp1/nativeapp1.csproj]
/Users/am11/.dotnet7/sdk/7.0.100-preview.7.22351.3/Roslyn/Microsoft.CSharp.Core.targets(75,5): error :    at Microsoft.CodeAnalysis.CSharp.Binder.CheckValEscape(Microsoft.CodeAnalysis.SyntaxNode, Microsoft.CodeAnalysis.CSharp.BoundExpression, UInt32, UInt32, Boolean, Microsoft.CodeAnalysis.CSharp.BindingDiagnosticBag) [/Users/am11/nativeapp1/nativeapp1.csproj]
/Users/am11/.dotnet7/sdk/7.0.100-preview.7.22351.3/Roslyn/Microsoft.CSharp.Core.targets(75,5): error : -------------------------------- [/Users/am11/nativeapp1/nativeapp1.csproj]
/Users/am11/.dotnet7/sdk/7.0.100-preview.7.22351.3/Roslyn/Microsoft.CSharp.Core.targets(75,5): error :    at Microsoft.CodeAnalysis.CSharp.Binder.ValidateEscape(Microsoft.CodeAnalysis.CSharp.BoundExpression, UInt32, Boolean, Microsoft.CodeAnalysis.CSharp.BindingDiagnosticBag) [/Users/am11/nativeapp1/nativeapp1.csproj]
/Users/am11/.dotnet7/sdk/7.0.100-preview.7.22351.3/Roslyn/Microsoft.CSharp.Core.targets(75,5): error :    at Microsoft.CodeAnalysis.CSharp.Binder.<BindExpressionBodyAsBlock>g__bindExpressionBodyAsBlockInternal|960_0(Microsoft.CodeAnalysis.CSharp.Syntax.ArrowExpressionClauseSyntax, Microsoft.CodeAnalysis.CSharp.Binder, Microsoft.CodeAnalysis.CSharp.BindingDiagnosticBag) [/Users/am11/nativeapp1/nativeapp1.csproj]
/Users/am11/.dotnet7/sdk/7.0.100-preview.7.22351.3/Roslyn/Microsoft.CSharp.Core.targets(75,5): error :    at Microsoft.CodeAnalysis.CSharp.MethodCompiler.BindMethodBody(Microsoft.CodeAnalysis.CSharp.Symbols.MethodSymbol, Microsoft.CodeAnalysis.CSharp.TypeCompilationState, Microsoft.CodeAnalysis.CSharp.BindingDiagnosticBag, Boolean, Microsoft.CodeAnalysis.CSharp.BoundNode, Boolean, Microsoft.CodeAnalysis.CSharp.ImportChain ByRef, Boolean ByRef, Boolean ByRef, InitialState ByRef) [/Users/am11/nativeapp1/nativeapp1.csproj]
/Users/am11/.dotnet7/sdk/7.0.100-preview.7.22351.3/Roslyn/Microsoft.CSharp.Core.targets(75,5): error :    at Microsoft.CodeAnalysis.CSharp.MethodCompiler.CompileMethod(Microsoft.CodeAnalysis.CSharp.Symbols.MethodSymbol, Int32, ProcessedFieldInitializers ByRef, Microsoft.CodeAnalysis.CSharp.SynthesizedSubmissionFields, Microsoft.CodeAnalysis.CSharp.TypeCompilationState) [/Users/am11/nativeapp1/nativeapp1.csproj]
/Users/am11/.dotnet7/sdk/7.0.100-preview.7.22351.3/Roslyn/Microsoft.CSharp.Core.targets(75,5): error :    at Microsoft.CodeAnalysis.CSharp.MethodCompiler.CompileNamedType(Microsoft.CodeAnalysis.CSharp.Symbols.NamedTypeSymbol) [/Users/am11/nativeapp1/nativeapp1.csproj]
/Users/am11/.dotnet7/sdk/7.0.100-preview.7.22351.3/Roslyn/Microsoft.CSharp.Core.targets(75,5): error :    at Microsoft.CodeAnalysis.CSharp.MethodCompiler+<>c__DisplayClass25_0.<CompileNamedTypeAsync>b__0() [/Users/am11/nativeapp1/nativeapp1.csproj]
/Users/am11/.dotnet7/sdk/7.0.100-preview.7.22351.3/Roslyn/Microsoft.CSharp.Core.targets(75,5): error :    at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(System.Threading.Thread, System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object) [/Users/am11/nativeapp1/nativeapp1.csproj]
/Users/am11/.dotnet7/sdk/7.0.100-preview.7.22351.3/Roslyn/Microsoft.CSharp.Core.targets(75,5): error :    at System.Threading.Tasks.Task.ExecuteWithThreadLocal(System.Threading.Tasks.Task ByRef, System.Threading.Thread) [/Users/am11/nativeapp1/nativeapp1.csproj]
/Users/am11/.dotnet7/sdk/7.0.100-preview.7.22351.3/Roslyn/Microsoft.CSharp.Core.targets(75,5): error :    at System.Threading.ThreadPoolWorkQueue.Dispatch() [/Users/am11/nativeapp1/nativeapp1.csproj]
/Users/am11/.dotnet7/sdk/7.0.100-preview.7.22351.3/Roslyn/Microsoft.CSharp.Core.targets(75,5): error :    at System.Threading.PortableThreadPool+WorkerThread.WorkerThreadStart() [/Users/am11/nativeapp1/nativeapp1.csproj]
/Users/am11/.dotnet7/sdk/7.0.100-preview.7.22351.3/Roslyn/Microsoft.CSharp.Core.targets(75,5): error :    at System.Threading.Thread.StartCallback() [/Users/am11/nativeapp1/nativeapp1.csproj]

The build failed. Fix the build errors and run again.
```

**Additional info**

Without concatenation, the compilation of (combined) u8 literal with length even > 10K does not overflow the stack. So it is the problem with recently added u8 literal concatenation in v4.4.0-1.22328.22.

**Comment by am11** [comment]
Normal strings literals do not have this problem when concatenated. Can ""u8 concatenation use the same mechanism as the normal string? (run validation after joining the whole thing)

**Comment by jcouv** [comment]
Fixed merged for 17.4
